File: backend/app/schemas/interpret.py
"""Pydantic-scheman för AI-tolkning."""
from typing import Literal

from pydantic import BaseModel, Field


InterpretKind = Literal[
    "dream",
    "poem",
    "reflection",
    "text_excerpt",
    "symbol_focus",
    "event_experience",
    "relation_situation",
    "free",
]
CautionLevel = Literal["high", "medium"]

INTERPRET_KIND_VALUES: frozenset[str] = frozenset(
    (
        "dream",
        "poem",
        "reflection",
        "text_excerpt",
        "symbol_focus",
        "event_experience",
        "relation_situation",
        "free",
    )
)


class InterpretRequestBody(BaseModel):
    """Kropp för POST /posts/{id}/interpret (query-param `model` stöds fortfarande som alternativ)."""

    model: str | None = None
    interpret_kind: InterpretKind | None = None


class SupportedModel(BaseModel):
    """Modell som backend kan erbjuda till klienten."""

    id: str
    label: str
    description: str
    runtime_available: bool = Field(
        default=True,
        description="False om runtime-listan sades nej till just denna modell-id.",
    )


class InterpretationSection(BaseModel):
    """En strukturerad del i AI-svaret."""

    id: str
    title: str
    content: str


class InterpretationContract(BaseModel):
    """Sammanfattning av vilket tolkningskontrakt som användes."""

    kind: InterpretKind
    label: str
    tone: str
    caution_level: CautionLevel
    section_titles: list[str]


class InterpretResponse(BaseModel):
    """Strukturerat AI-svar för en post."""

    interpretation: str
    model_used: str
    requested_model: str | None = None
    used_model: str
    fallback_used: bool = False
    fallback_reason: str | None = None
    provider: str = "openai"
    contract_degraded: bool = False
    disclaimer: str
    contract: InterpretationContract
    sections: list[InterpretationSection]


class InterpretStatus(BaseModel):
    """Status och modellstöd för AI-funktionen."""

    available: bool
    model_default: str
    model_options: list[SupportedModel]
    runtime_verification_succeeded: bool = False
    runtime_verification_message: str | None = None

File: frontend/src/services/interpret.ts
import { get, post } from "./api";

export type InterpretKind =
  | "dream"
  | "poem"
  | "reflection"
  | "text_excerpt"
  | "symbol_focus"
  | "event_experience"
  | "relation_situation"
  | "free";

export interface InterpretModelOption {
  id: string;
  label: string;
  description: string;
  runtime_available?: boolean;
}

export interface InterpretationSection {
  id: string;
  title: string;
  content: string;
}

export interface InterpretationContract {
  kind: InterpretKind;
  label: string;
  tone: string;
  caution_level: "high" | "medium";
  section_titles: string[];
}

export interface InterpretResponse {
  interpretation: string;
  model_used: string;
  requested_model: string | null;
  used_model: string;
  fallback_used: boolean;
  fallback_reason: string | null;
  provider: string;
  contract_degraded: boolean;
  disclaimer: string;
  contract: InterpretationContract;
  sections: InterpretationSection[];
}

export interface InterpretStatus {
  available: boolean;
  model_default: string;
  model_options: InterpretModelOption[];
  runtime_verification_succeeded: boolean;
  runtime_verification_message: string | null;
}

export function fetchInterpretStatus(): Promise<InterpretStatus> {
  return get<InterpretStatus>("/interpret/status");
}

export interface InterpretPostOptions {
  model?: string;
  interpret_kind?: InterpretKind;
}

export function interpretPost(postId: number, options?: InterpretPostOptions): Promise<InterpretResponse> {
  const body: Record<string, unknown> = {};
  if (options?.model) body.model = options.model;
  if (options?.interpret_kind) body.interpret_kind = options.interpret_kind;
  return post<InterpretResponse>(`/posts/${postId}/interpret`, body);
}
